### PR TITLE
ALE with finite volume for fluid flow

### PR DESCRIPTION
**Pull request by lindsayad**
## Reason
The AMMT program expects us to do ALE with finite volumes for the fluid flow and CFEM for the displacements

## Design
- Significant development to enable finite volume on displaced mesh/problem
- Add `NodeArg` for functor evaluation
- Implement `ElemArg` for FE for coupling the mesh velocity into the convective term of the momentum and energy transport equations
- Implement Qp evaluations for FV by just passing to `ElemPointArg` overload which does a two term expansion

## Impact
ALE simulation capability with finite volume modeling the NS equations and nodal displacements for the displacing mesh

## Results

This is a two-dimensional melt pool simulation. Coloring is temperature and arrows are pointing in direction of velocity and scaled by the velocity magnitude. A laser spot x-coordinate is a sinusoidal function of time, ratcheting back and forth in the middle half of the domain

https://github.com/idaholab/moose/assets/10248304/8fc770e1-5b16-46ea-875b-dc7793749cfe

### CODE REVIEW

**Comment by jiangwen84** [comment]
@lindsayad  This looks great. According to your experience, is there any apparent pros of FV version of ALE?

**Comment by lindsayad** [comment]
My experience thus far is that it's a lot harder to make it work 😅 It seems less robust; I've had a harder time getting good results for a given set of parameters. For instance, I haven't gotten good results yet in which the "melt pool" starts from a uniformly cold temperature, e.g. 300K, in which the entire medium is initially solid. I was really close to giving up before getting the result shown above. It was pure luck that I switched to an IC in which the material is liquid and observed that I could get some results. So I think there's still a lot to learn and explore

**Comment by jiangwen84** [comment]
> My experience thus far is that it's a lot harder to make it work 😅 It seems less robust; I've had a harder time getting good results for a given set of parameters. For instance, I haven't gotten good results yet in which the "melt pool" starts from a uniformly cold temperature, e.g. 300K, in which the entire medium is initially solid. I was really close to giving up before getting the result shown above. It was pure luck that I switched to an IC in which the material is liquid and observed that I could get some results. So I think there's still a lot to learn and explore

Haha, did not know it was that harder for you :-) I will explore it too. This is really a new and exciting capability. thanks!

File: framework/contrib/wasp
Comment (GiudGiud) [inline framework/contrib/wasp:1]: dont think this belongs

File: framework/include/fvkernels/FVPointValueConstraint.h
Comment (GiudGiud) [inline framework/include/fvkernels/FVPointValueConstraint.h:35]: pressurePin will probably want the same. Does it work better by any chance?

File: modules/navier_stokes/doc/content/source/kernels/INSADMomentumMeshAdvection.md
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/kernels/INSADMomentumMeshAdvection.md:6]:
```suggestion
the term representing the contribution of the mesh displacement to the conservation of momentum:
```

Comment (GiudGiud) [reply modules/navier_stokes/doc/content/source/kernels/INSADMomentumMeshAdvection.md:6]: should there be an INSADMassMeshAdvection? seems like moving the mesh moves mass

File: modules/heat_conduction/include/bcs/GaussianEnergyFluxBC.h
Comment (GiudGiud) [inline modules/heat_conduction/include/bcs/GaussianEnergyFluxBC.h:26]: docstring

**Comment by GiudGiud** [comment]
> Implement Qp evaluations for FV by just passing to ElemPointArg overload which does a two term expansion
 
Is this for elem Qp or sideQp ?

File: framework/include/variables/MooseVariableFE.h
Comment (GiudGiud) [inline framework/include/variables/MooseVariableFE.h:775]: I kind of wish there was a way to put this code in a separate class, so all the functor stuff lives together.

But a base class would need to still be in the current chain of variable classes to have all the attributes needed
and a derived class would be troublesome in its own way

File: framework/include/variables/MooseVariableFV.h
Comment (GiudGiud) [inline framework/include/variables/MooseVariableFV.h:729]: this is not that intruitie right so let's make it clear in the docs

File: framework/src/systems/NonlinearSystemBase.C
Comment (GiudGiud) [inline framework/src/systems/NonlinearSystemBase.C:1557]: add a comment?
we need to do a face-info loop once for undisplaced objects and once for diplaced ones

File: modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h
Comment (GiudGiud) [inline modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h:50]: does this check the block restrictions too

File: modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h
Comment (GiudGiud) [inline modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h:57]: should we check block restriction of that one vs this kernel

File: modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C
Comment (GiudGiud) [inline modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C:25]: this could be limiting in the futue, add a node to split the nodeset if the subdomain dont extend to the entire nodeset?

File: modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C
Comment (GiudGiud) [inline modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C:49]: hardcoding Euler

lets have some check in the constructor?

**Review by GiudGiud — COMMENTED** [review]
looked over the whole thing. 

it would be good to see a reference solution being matched, for example something pretty contrived in 1D where we impose a displacement and watch the incompressible fluid follow?

or MMS?

File: framework/include/fvkernels/FVPointValueConstraint.h
Comment (lindsayad) [reply framework/include/fvkernels/FVPointValueConstraint.h:35]: You look for a new element every time you `execute` so `NSFVPressurePin` seems fine as is.

We don't really have any tests of this class with adaptivity, so my only reference point is the new laser welding input ... and yes with this change adaptivity definitely works better ... the singular row is no longer singular

File: framework/include/variables/MooseVariableFE.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableFE.h:775]: This kind of caching is unique to the reinit'ing of FE objects so to me this is an ideal place for it

File: framework/src/variables/MooseVariableFV.C
Comment (lindsayad) [reply framework/src/variables/MooseVariableFV.C:54]: I can change the comment to reflect the fact that there is not a single FV input in our repository that does not require a layer of ghosting

File: modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h
Comment (lindsayad) [reply modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h:50]: No but the Problem does integrity checks for that

File: modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h
Comment (lindsayad) [reply modules/navier_stokes/include/kernels/INSADMomentumMeshAdvection.h:57]: The tracker is a general user object so it's not block restrictable. It bases its "block restriction" based off the blocks of the INSADMaterials. I already did have an error for this but the message was not very good. I will improve it

File: modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C
Comment (lindsayad) [reply modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C:25]: I'm not quite sure what you mean in your second clause

File: modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C
Comment (lindsayad) [reply modules/navier_stokes/src/bcs/INSADDisplaceBoundaryBC.C:49]: It's already there 10 lines up per your suggestion on the original PR that added this object :-)

File: modules/navier_stokes/src/fvkernels/INSFVAdvectionKernel.C
Comment (lindsayad) [reply modules/navier_stokes/src/fvkernels/INSFVAdvectionKernel.C:37]: This is a remnant of earlier. I've removed it. I will add some new checking for consistency between 'use_displaced_mesh' and displacement parameters in the interpolator object

File: framework/src/variables/MooseVariableFE.C
Comment (lindsayad) [reply framework/src/variables/MooseVariableFE.C:934]: For std::vectors, this would only add zeros for new elements

File: framework/src/variables/MooseVariableFE.C
Comment (lindsayad) [reply framework/src/variables/MooseVariableFE.C:941]: This is needed to ensure that all qp-entries are zeroed. The suggestion above would only zero new elements

File: framework/src/variables/MooseVariableFE.C
Comment (lindsayad) [reply framework/src/variables/MooseVariableFE.C:1127]: I'll add the comment that I added for the elem-qp version

**Comment by lindsayad** [comment]
> for example something pretty contrived in 1D where we impose a displacement and watch the incompressible fluid follow?
> 
> or MMS?

I've added an MMS case that tests the incorporation of the mesh velocity in the advecting velocity

File: framework/src/variables/MooseVariableFE.C
Comment (GiudGiud) [inline framework/src/variables/MooseVariableFE.C:846]: we could also assert that this dof be a value dof as opposed to a gradient dof

and/or that there be only one dof on the node (per component if vector)

File: framework/src/postprocessors/VectorSum.C
Comment (lindsayad) [reply framework/src/postprocessors/VectorSum.C:33]: The communication happens in the `sum()` call on the vector

File: modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C
Comment (lindsayad) [reply modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C:47]: I amended the doc string for `cp` to say it should be a constant. I do not know how to integrate some arbitrary expression for `_cp` here in this functor. In the Aria material since I have an analytic expression for cp, I implement the analytic integral for cp dtau for h

Comment (GiudGiud) [reply modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C:47]: oh yeah my bad. 

We should just query _fp directly for h instread of _cp, we should not try to integrate ourselves

File: framework/src/variables/MooseVariableFE.C
Comment (lindsayad) [reply framework/src/variables/MooseVariableFE.C:846]: I don't know what a gradient dof is. I don't know of finite element approximate formulations other than u = sum(i) phi_i * u_i

File: modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C
Comment (GiudGiud) [reply modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C:47]: doc string update is fine

File: framework/src/variables/MooseVariableFE.C
Comment (GiudGiud) [reply framework/src/variables/MooseVariableFE.C:846]: I think it s a thing in our exotic families, like szabab. This is the kind of thing I just ask Roy about tbh

We essentially do not support them in MOOSE in both the auxkernel and transfer systems

**Review by GiudGiud — DISMISSED** [review]
functor unit test could use a freshener if you want to get framework coverage up. It s low because the code is tested in modules.

otherwise the new results look good, the mms order in the tests are a good sign and I don't see any issue in the code.

File: modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolatorSegregated.C
Comment (GiudGiud) [inline modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolatorSegregated.C:159]: should you assert on subtract_mesh_velocity since it s not implemented

Comment (lindsayad) [reply modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolatorSegregated.C:159]: I think I will just error if this object is `_displaced`. It's valid for a "general" postprocessor to call this method with `true` as long as the calculation is not actually displaced

**Comment by lindsayad** [comment]
Ok added a meaningful test that should boost framework coverage

File: modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C
Comment (GiudGiud) [reply modules/navier_stokes/src/materials/INSFVEnthalpyMaterial.C]: umm it was specific

**Review by GiudGiud — DISMISSED** [review]

**Comment by lindsayad** [comment]
You'll have to re-approve after the heat transfer merge haha

**Review by GiudGiud — DISMISSED** [review]

**Review by GiudGiud — APPROVED** [review]

**Comment by lindsayad** [comment]
All but the Pronghorn failures are unrelated and there's a patch up for that

**Comment by GiudGiud** [comment]
Thanks @AlbertJP7 for your contribution on nodal functor arguments! They are actually going to be really important for coupling 1D physics with 3D physics at nodes.

And congrats @lindsayad  on getting this ALE capability in!

**Comment by lindsayad** [comment]
Thanks, and thank you for your review as always!

**Pull request merged**
